### ISSUE TITLE
Alternator: rack-aware RF heuristic

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5092]
Our current heuristics for new tables' RF (determined when the first table is created - as its keyspace is created), it the following: We always use RF=3 - unless there are fewer than 3 nodes (in which case we use RF=1).

But if the cluster has **5 racks** in each DC, for example, it was likely the intention of the person creating this cluster that RF=5 be used. So we can consider the number of racks in the heuristic that determines RF.

**Comment by slivne** [comment]
The other alternative is that the user has selected 5 for capacity - I think the default of 3 is ok

**Comment by nyh** [comment]
@noamha can you please comment on @slivne 's comment? In your experience, do we sometimes use more racks (e.g., AWS availablity zones) than the requested RF?

**Comment by tzach** [comment]
https://github.com/scylladb/scylla/issues/4412

**Comment by noamha** [comment]
@nyh It's less popular but we do have a cluster with RF of 4 in production.

**Comment by nyh** [comment]
> The other alternative is that the user has selected 5 for capacity - I think the default of 3 is ok

I saw a user who wanted to use 5 racks specifically so that each rack will have a working replica and will be able to locally serve clients on the same rack, so this user would want RF=5 to be used, not RF=3.

If we want to allow both use cases, we can pick a default (Perhaps "RF is number of racks"?) and allow the user to override the RF using a tag in CreateTable.

CC @ScyllaPiotr

**Comment by nyh** [comment]
I'm considering not changing the existing default - but rather make this a per-table-creation option:

1. new tables will continue to use RF=3 regardless of number of racks
2. A tag `system:replication_factor` can be set on the new table; Setting it to '5' will create the table with RF=5.

This won't be exactly what I asked for when creating this issue - it won't be a heuristic, it will be an option the user will need to explicitly request.

Opinions anyone?

**Comment by ScyllaPiotr** [comment]
> I'm considering not changing the existing default - but rather make this a per-table-creation option:
>   1. new tables will continue to use RF=3 regardless of number of racks
>   2. A tag `system:replication_factor` can be set on the new table; Setting it to '5' will create the table with RF=5.
> 
> This won't be exactly what I asked for when creating this issue - it won't be a heuristic, it will be an option the user will need to explicitly request.
> 
> Opinions anyone?

I like this. There can be clusters with 5 or 9 nodes to provide (allow for) higher performance and capacity, not to increase redundancy. Smaller nodes make for changing cluster topology easier.
So yes, I like your latest idea the most.

**Comment by ScyllaPiotr** [comment]
Can we change RF for an existing ks? If so, let's support it with `UpdateTable` too.

**Comment by nyh** [comment]
> Can we change RF for an existing ks? If so, let's support it with `UpdateTable` too.

I don't know how, let's ask the real experts (@tgrabiec ?). I assume that changing the replication factor on an existing table requires some sort of lengthy rebalancing of the data, and I don't know if it's even allowed.

In any case, since I propose setting the table's replication factor by setting a "tag" on the table, the interface to change it - if we ever allow it - will be to write to the tag, not by `UpdateTable`. Right now my plan is that writing this tag on an existing table will have no effect (and the documentation will explain it).

**Comment by nyh** [comment]
>> Can we change RF for an existing ks?
> I don't know how, let's ask the real experts

Even faster is asking Claude. It answered the following, I hope it's not an hallucination. 
The good news (if it's true) is that **with tablets, we should allow changing the RF on-the-fly** using the API and no weird CQL or external hacks.

Here is the full reply:

You can use `ALTER KEYSPACE`.

This changes the RF in the keyspace metadata immediately. With **vnodes**, the new replica nodes don't automatically receive the existing data — you need to run nodetool repair on all nodes afterward to stream the missing data to the new replicas.

However, with **tablets** (the newer replication mode), the tablet balancer handles redistribution automatically, so repair is not needed — the balancer will migrate tablet replicas to satisfy the new RF target on its own.

Reads and writes continue working normally throughout — the rebalancing is fully online and non-blocking.

Writes go to however many replicas currently exist for each tablet. Once a new replica is added to a tablet (after migration completes for that tablet), future writes go to all RF replicas.

Reads use whatever replicas are currently available. If a read uses consistency level QUORUM, it only needs a quorum of the current RF, so it will succeed using the old replicas while the balancer is still working.

The key point is that tablet migration is done tablet-by-tablet, not all at once. Each tablet independently transitions to the new RF: the balancer streams the data to the new replica node, and once that specific tablet's migration is complete and acknowledged, that tablet operates at the new RF. Other tablets may still be at the old RF. The table remains fully available during the entire process.

**Comment by nyh** [comment]
> **with tablets, we should allow changing the RF on-the-fly** using the API and no weird CQL or external hacks.

The DynamoDB API to do this would be to write to the tag `system:replication_factor` tag.  However, to make this work we'll need to make some changes to `executor::tag_resource()` 
1. If the `system:replication_factor` tag is changed, a keyspace modification to change the RF is performed before or after changing the tags.
2. It will be better - but more complicated - to do the keyspace change and tag change atomically (on can't succeed without the other). To do this we'll need to stop using the convenient `db::modify_tags()` utility function, and instead need to open-code it as we did in UpdateTable: take the group 0 lock, prepare both the keyspace change and the tag change, and finally announce() it together.

We also need to consider what to do with untag_resource() of this tag. Will it be forbidden? Or maybe removing the tag should move us to the default RF (3)?

By the way: Changing the RF of a table is a very serious operation. Increasing RF can suddenly use a lot more disk space, and decreasing RF can reduce replication. Wouldn't it be a bit scary that a benign-looking tag modification can do it?

**Comment by ScyllaPiotr** [comment]
> We also need to consider what to do with untag_resource() of this tag. Will it be forbidden? Or maybe removing the tag should move us to the default RF (3)?

The latter, no?

> Changing the RF of a table is a very serious operation. Increasing RF can suddenly use a lot more disk space, and decreasing RF can reduce replication. Wouldn't it be a bit scary that a benign-looking tag modification can do it?

Is there a better way for changing the RF for an existing table?

**Comment by tgrabiec** [comment]
> > Can we change RF for an existing ks? If so, let's support it with `UpdateTable` too.
> 
> I don't know how, let's ask the real experts ([@tgrabiec](https://github.com/tgrabiec) ?). I assume that changing the replication factor on an existing table requires some sort of lengthy rebalancing of the data, and I don't know if it's even allowed.

Yes, RF can be changed using ALTER. Expansion is expensive, it waits for new replicas to be rebuilt.
